arca/backend/base.py

```diff
@@ -95,11 +104,46 @@ def serialized_task(self, task: Task) -> Tuple[str, str]:
         return f"{task.hash}.json", task.json
 
     def get_requirements_hash(self, requirements_file: Path) -> str:
-        """ Returns an SHA1 hash of the contents of the ``requirements_path``.
+        """ Returns an SHA256 hash of the contents of the ``requirements_path``.
         """
         logger.debug("Hashing: %s%s", requirements_file.read_text(), arca.__version__)
         return hashlib.sha256(bytes(requirements_file.read_text() + arca.__version__, "utf-8")).hexdigest()
 
+    def get_pipfiles(self, path: Path) -> PipfilesType:
+        """
+        Checks the repository if it contains Pipenv files, if not returns ``None`` right away.
+        Otherwise it returns a tuple which contain :class:`Path <pathlib.Path>` instances to the files.
+        If ``Pipfile`` is in the repository but ``Pipfile.lock`` is missing then
+        the second item in the tuple is ``None``.
+        """
+        if self.pipfile_location is None:
+            return None
+
+        pipfile = path / self.pipfile_location / "Pipfile"
+        pipfile_lock = path / self.pipfile_location / "Pipfile.lock"
+
+        if not pipfile.exists():
+            return None
+
+        return pipfile, pipfile_lock if pipfile_lock.exists() else None
+
+    def get_pipfile_hash(self, pipfile: Path, pipfile_lock: Optional[Path]) -> str:
+        """
+        Returns a SHA256 hash of the Pipenv file that is gonna be used for install.
+        (Lock file if present, otherwise Pipfile.)
+        """
+        if pipfile_lock is not None:
+            pipfile_lock_contents = pipfile_lock.read_text("utf-8")
+
+            try:
+                to_hash = json.loads(pipfile_lock_contents)["_meta"]["hash"]["sha256"]
```

Comment:
This looks a bit problematic: it gets the hash from the file, i.e. from user input.
Could a malicious user poison the cache using a well-crafted hash?
Could we check here that the hash actually matches the lockfile?

Reply:
Alternatively, hash the *contents* of the lockfile.
Either hash the file itself, or some normalized version of the contents so that whitespace doesn't matter.
(AFAICS, pipenv [normalizes data by dumping it as JSON](https://github.com/pypa/pipfile/blob/master/pipfile/api.py#L117), and looks like it's getting away with it.)